### ISSUE TITLE
1.0 ng1: Hooks can still be called after de-register invocation 

### ISSUE THREAD

**Issue by jgrund** [angular-ui/ui-router#2928]
We are seeing a case where our hook callback is invoked after de-registering it.

It seems that the callback should never be invoked after a de-register is called.

**Comment by johnsonw** [comment]
👍 , seeing the same thing.

**Comment by christopherthielen** [comment]
Are you deregistering the hook while a transition is in process? The hook pipeline is computed when the transition starts. 

What is your use case(s)?

**Comment by jgrund** [comment]
So the basic scenario can be boiled down to:

```
   function serviceWithTransition ($transitions) {
     return () => 
       $transitions.onSuccess({}, console.log);
   };

  const component = {
    controller: function (serviceWithTransition) {
      this.$onDestroy = serviceWithTransition;
    },
    template: '<foo></foo>'
  }
```

Where `component` is registered as a component on a state.

What ends up happening is: 

1. `$onDestroy` is invoked on a route transition.
2. The `onSuccess` handler is de-registered.
3. The `onSuccess` callback fires for the component being navigated away from, even though it was de-registered previously.

What I considered surprising about this is that the de-registration function is not upheld 100 percent of the time.

It seems like this can lead to subtle bugs as I suspect others with have the same assumption.

**Comment by christopherthielen** [comment]
I see... that's helpful, thanks for the use case details.

I agree that's surprising behavior as you described.

---

Here's the details of why it's happening:

1) Transition to first state
2) Component is created, hook is registered in global registry
3) Transition away from first state
  - onBefore hooks are fetched and sorted by priority, then run in order
  - all matching (async) hooks are fetched from global registry
  - those hooks are and sorted by hook type/hook priority, then queued for run in order
  - Transition is successful.
4) onSuccess hooks are fetched and sorted by priority
  - your `onSuccess` hook is placed in the onSuccess pipeline at this time
  - Views are updated
    - The component is destroyed, hook deregister function is invoked
    - The pipeline is unaffected

---

I'm going to have to give some thought to this.

**Comment by christopherthielen** [comment]
Fixed by #2939
